# Описание проекта
Вы решили открыть небольшое кафе в Москве. Оно оригинальное — гостей должны обслуживать роботы. Проект многообещающий, но дорогой. Вместе с партнёрами вы решились обратиться к инвесторам. Их интересует текущее положение дел на рынке — сможете ли вы снискать популярность на долгое время, когда все зеваки насмотрятся на роботов-официантов?
Вы — гуру аналитики, и партнёры просят вас подготовить исследование рынка. У вас есть открытые данные о заведениях общественного питания в Москве.

<br>
<a name="up">Содержание</a>
<br>
<br>


<a href="#step_one"><h4>Шаг №1</h4><p>Предобработка данных</p></a>

<h4><a href="#step_two">Шаг №2</a></h4>

<ol>
    <li>
        <a href="#s2_1_1">Исследуйте соотношение видов объектов общественного питания по количеству.</a>
        <a href="#s2_1_2">Постройте график.</a> 
        <a href="#s2_1_3">Вывод</a>
    </li>
        <br>
    <li>
        <a href="#s2_2_1">Исследуйте соотношение сетевых и несетевых заведений по количеству.</a>
        <a href="#s2_2_2">Постройте график.</a> 
    </li>
        <br>
    <li>
        <a href="#s2_3">Для какого вида объекта общественного питания характерно сетевое распространение?</a>
    </li>
        <br>
    <li>
        <a href="#s2_4">Что характерно для сетевых заведений: много заведений с небольшим числом посадочных мест в каждом или мало заведений с большим количеством посадочных мест?</a>
    </li>
        <br>
    <li>
        <a href="#s2_5">Для каждого вида объекта общественного питания опишите среднее количество посадочных мест. Какой вид предоставляет в среднем самое большое количество посадочных мест? Постройте графики.</a>
    </li>
        <br>
    <li>
        <a href="#s2_6">Выделите в отдельный столбец информацию об улице из столбца address</a>
    </li>
        <br>
    <li>
        <a href="#s2_7_1">Постройте график топ-10 улиц по количеству объектов общественного питания. Воспользуйтесь внешней информацией и ответьте на вопрос — в каких районах Москвы находятся эти улицы?</a>
        <a href="#s2_7_2">Вывод.</a>
        <a href="#s2_7_3">Внешние источники и ссылки.</a>
    </li>
        <br>
    <li>
        <a href="#s2_8_1">Найдите число улиц с одним объектом общественного питания. Воспользуйтесь внешней информацией и ответьте на вопрос — в каких районах Москвы находятся эти улицы?</a>
        <a href="#s2_8_2">Вывод.</a>
        <br>
    <li>
        <a href="#s2_9_1">Посмотрите на распределение количества посадочных мест для улиц с большим количеством объектов общественного питания. Какие закономерности можно выявить?</a>
        <a href="#s2_9_2">Вывод.</a>
    </li>
</ol>

<a href="#s2_10">Общий вывод и рекомендации о виде заведения, количестве посадочных мест, а также районе расположения. Возможность развития сети.</a>

In [1]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import numpy as np
import re
import random
import json
import requests

---
<a href="#up">Наверх</a>
<a name="step_one"><h2>Шаг №1 - предобработка данных</h2></a>

In [2]:
data = pd.read_csv('rest_data.csv')
# Выведем таблицу на экран:
display(data)
print()
display(data.info())

id             object_name chain object_type  \
0      151635                 СМЕТАНА   нет        кафе   
1       77874                  Родник   нет        кафе   
2       24309         Кафе «Академия»   нет        кафе   
3       21894              ПИЦЦЕТОРИЯ    да        кафе   
4      119365  Кафе «Вишневая метель»   нет        кафе   
...       ...                     ...   ...         ...   
15361  208537             Шоколадница    да        кафе   
15362  209264             Шоколадница    да        кафе   
15363  209186             Шоколадница    да        кафе   
15364  221900             Шоколадница    да        кафе   
15365  222535             Шоколадница    да        кафе   

                                                 address  number  
0             город Москва, улица Егора Абакумова, дом 9      48  
1      город Москва, улица Талалихина, дом 2/1, корпус 1      35  
2              город Москва, Абельмановская улица, дом 6      95  
3                город Москва, Абрамцевская улица, дом 1      40  
4      город Москва, Абрамцевская улица, дом 9, корпус 1      50  
...                                                  ...     ...  
15361       город Москва, 3-й Крутицкий переулок, дом 18      50  
15362           город Москва, улица Земляной Вал, дом 33      10  
15363           город Москва, улица Земляной Вал, дом 33      20  
15364  город Москва, поселение Московский, Киевское ш...      36  
15365             город Москва, Ходынский бульвар, дом 4      10  

[15366 rows x 6 columns]


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15366 entries, 0 to 15365
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           15366 non-null  int64 
 1   object_name  15366 non-null  object
 2   chain        15366 non-null  object
 3   object_type  15366 non-null  object
 4   address      15366 non-null  object
 5   number       15366 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 720.4+ KB


None

In [3]:
# Проверим на дубликаты:
data.duplicated().sum()

0

In [4]:
# Посмотрим сколько сетевых заведений в таблице:
len(data['object_name'].unique())

10393

In [5]:
# Посмотрим действительно ли в столбце chain содержится всего 2 переменные:
len(data['chain'].unique())

2

In [6]:
# Изучим какие заведения приведены в датасете:
data['object_type'].unique()

array(['кафе', 'столовая', 'закусочная',
       'предприятие быстрого обслуживания', 'ресторан', 'кафетерий',
       'буфет', 'бар', 'магазин (отдел кулинарии)'], dtype=object)

---
<a href="#up">Наверх</a>
<a name="step_two"><h2>Шаг №2 - анализ данных</h2></a>

---
<a href="#up">Наверх</a>
<a name="s2_1_1"></a>
### Соотношение видов объектов общественного питания по количеству

In [7]:
data_agg = data.groupby('object_type').agg({'id':'count'})
data_agg.rename(columns={'id':'number_of_rest'}, inplace=True)
data_agg['percents'] = data_agg['number_of_rest'] / data_agg['number_of_rest'].sum()
data_agg.round(3)

number_of_rest  percents
object_type                                                
бар                                           856     0.056
буфет                                         585     0.038
закусочная                                    360     0.023
кафе                                         6099     0.397
кафетерий                                     398     0.026
магазин (отдел кулинарии)                     273     0.018
предприятие быстрого обслуживания            1923     0.125
ресторан                                     2285     0.149
столовая                                     2587     0.168

Посмотрим какие заведения являются самыми малочисленными и посчитаем суммарную долю всех заведений, доля которых составляет меньше 10% от общего количества заведений:

In [8]:
print(f"Суммарная доля - {data_agg[data_agg['percents'] < 0.1]['percents'].sum().round(3)*100}%")

Суммарная доля - 16.1%


---
<a href="#up">Наверх</a>
<a name="s2_1_2"></a>
### График соотношения видов объектов общественного питания по количеству

In [9]:
data

id             object_name chain object_type  \
0      151635                 СМЕТАНА   нет        кафе   
1       77874                  Родник   нет        кафе   
2       24309         Кафе «Академия»   нет        кафе   
3       21894              ПИЦЦЕТОРИЯ    да        кафе   
4      119365  Кафе «Вишневая метель»   нет        кафе   
...       ...                     ...   ...         ...   
15361  208537             Шоколадница    да        кафе   
15362  209264             Шоколадница    да        кафе   
15363  209186             Шоколадница    да        кафе   
15364  221900             Шоколадница    да        кафе   
15365  222535             Шоколадница    да        кафе   

                                                 address  number  
0             город Москва, улица Егора Абакумова, дом 9      48  
1      город Москва, улица Талалихина, дом 2/1, корпус 1      35  
2              город Москва, Абельмановская улица, дом 6      95  
3                город Москва, Абрамцевская улица, дом 1      40  
4      город Москва, Абрамцевская улица, дом 9, корпус 1      50  
...                                                  ...     ...  
15361       город Москва, 3-й Крутицкий переулок, дом 18      50  
15362           город Москва, улица Земляной Вал, дом 33      10  
15363           город Москва, улица Земляной Вал, дом 33      20  
15364  город Москва, поселение Московский, Киевское ш...      36  
15365             город Москва, Ходынский бульвар, дом 4      10  

[15366 rows x 6 columns]

In [37]:
# Построим график:
obj_type = px.histogram(data, x='object_type')
obj_type.update_layout(
    title='График соотношения видов объектов общественного питания по количеству',
    xaxis={'categoryorder':'total descending'}, 
    xaxis_title_text='',
    yaxis_title_text='Количество объектов',
    template='plotly_dark'
    )

obj_type.show()

In [38]:
object_type_graph = go.Figure(data=[
    go.Pie(labels=data_agg.index, values=data_agg['number_of_rest'], hole=.2, textinfo='label+percent')
    ])
object_type_graph.update_layout(
    title='Доли предприятий общественного питания',    
    template='plotly_dark')
object_type_graph.show()

<a href="#up">Наверх</a>
<a name="s2_1_3"></a>
### Вывод
Самый распространённый вид заведений - это кафе, столовые, рестораны и кафе, а доля остальных – баров, буфетов, кафетериев, закусочных и отделов кулинарии – составляет 16%. 

---
<a href="#up">Наверх</a>
<a name="s2_2_1"></a>
### Соотношение сетевых и несетевых заведений по количеству

In [14]:
total_rest = data['chain'].count()
chain_rest = (data['chain'] == 'да').sum() 
non_chain_rest = (data['chain'] == 'нет').sum()
print(f'Общее количество заведений - {total_rest}. \
Из них сетевые – {(chain_rest/total_rest*100).round(2)}%, несетевые – {(non_chain_rest/total_rest*100).round(2)}%')

Общее количество заведений - 15366. Из них сетевые – 19.32%, несетевые – 80.68%


---
<a href="#up">Наверх</a>
<a name="s2_2_2"></a>
### График соотношения сетевых и несетевых заведений по количеству

In [15]:
total_rest_t = data.groupby('chain', as_index=False).agg({'id':'count'})
total_rest_labels = total_rest_t['chain']
total_rest_values = total_rest_t['id']

total_rest_graph = go.Figure(data=[go.Pie(
    values=total_rest_values, 
    labels=total_rest_labels, 
    textinfo='label+percent')])

total_rest_graph.update_layout(
    title='График соотношения сетевых и несетевых заведений',
    template='plotly_dark'
    )

total_rest_graph.show()

In [16]:
print(f"Процент сетевых заведений от общего количества предприятий быстрого обслуживания: {((data[(data['object_type'] == 'предприятие быстрого обслуживания') & (data['chain'] == 'да')]['id'].count()) / (data[data['object_type'] == 'предприятие быстрого обслуживания']['id'].count()) * 100).round(2)}%")

Процент сетевых заведений от общего количества предприятий быстрого обслуживания: 41.13%


In [17]:
total_rest_graph_obj_type = px.histogram(data, x='object_type', color='chain')

total_rest_graph_obj_type.update_layout(
    title_text='График количества сетевых и несетевых заведений по типу',
    xaxis_title_text='',
    yaxis_title_text='Количество заведений',
    bargroupgap=0,
    bargap=0.4,
    barmode='stack',
    template='plotly_dark'
)

total_rest_graph_obj_type.show()

<a href="#up">Наверх</a>
<a name="s2_3"></a>
### Вывод - для какого типа предприятия характерно сетевое распространение
#### Для предприятий быстрого обслуживания
Исходя из датасета, кафе, предприятия быстрого обслуживания и рестораны в больше степени являются сетевыми заведениями. Общее количество заведений - 15366. Из них сетевые – 19%, несетевые – 81%. Интересно при этом, что предприятия быстрого питания имеют самую большую долю сетевых заведений - 41%.

---
<a href="#up">Наверх</a>
<a name="s2_4"></a>
### Количество посадочных мест
Что характерно для сетевых заведений: много заведений с небольшим числом посадочных мест в каждом или мало заведений с большим количеством посадочных мест?

---
Вполне возможно, что количество посадочных мест в больших и в маленьких городах может различаться. Добавим город в новый датасет:

In [18]:
# Вполне возможно, что количество посадочных мест в больших и в маленьких городах может различаться. Добавим город:
data_with_city = data.copy()
data_with_city['city'] = data_with_city.apply(lambda x: x['address'].split(',')[0], axis=1)

In [19]:
data_with_city[data_with_city['city'] == 'город Москва'].shape[0]

15295

__Почти__ все значения - это Москва. Посмотим, что же не Москва:

In [20]:
print(f"Количество строк которые не содержат \"город Москва\" в адресе: {data_with_city[data_with_city['city'] != 'город Москва']['city'].shape[0]}")
print()
print(f"Из них уникальных: {len(data_with_city[data_with_city['city'] != 'город Москва']['city'].unique())}")
data_with_city[data_with_city['city'] != 'город Москва']['city'].unique()

Количество строк которые не содержат "город Москва" в адресе: 71

Из них уникальных: 54


array(['Партизанская улица', 'Поречная улица', '1-й Нагатинский проезд',
       'Софийская набережная', 'Нижний Таганский тупик',
       'Николоямская улица', 'улица Каманина', 'Погодинская улица',
       'Большая Бронная улица', '2-я Брестская улица',
       'Профсоюзная улица', 'Калашный переулок', 'Солянский тупик',
       'улица Земляной Вал', 'Таганская площадь', 'проезд Добролюбова',
       'улица Обручева', 'Вспольный переулок', 'улица Ленинская Слобода',
       'Дербеневская улица', 'Проектируемый проезд N 5231',
       'Ленинский проспект', 'Новоданиловская набережная',
       '5-й Котельнический переулок', 'улица Знаменка', 'Школьная улица',
       'поселение Сосенское', 'Полевая улица', 'поселение Марушкинское',
       'поселение Михайлово-Ярцевское', 'город Московский',
       'Митинская улица', 'Чертановская улица', 'Варшавское шоссе',
       'улица Миклухо-Маклая', '1-я Тверская-Ямская улица',
       'Краснобогатырская улица', 'город Троицк', 'Медынская улица',
       'Та

Итого 15295 московских адресов и 71 без указания адреса, но можно считать что это Москва, либо область.
По отсутствию названия города, а также по названиям улиц можно определить, что это либо Москва ([Проектируемый проезд N 5231](http://bit.ly/2TD64lq), [1-й Нагатинский проезд](http://bit.ly/2R7g70H), [1-я Тверская-Ямская улица](https://yandex.ru/maps/-/CKA0jQlP) и тд.), либо пригород ([город Троицк](https://yandex.kz/maps/-/CKA0jF3S), [поселение Михайлово-Ярцевское](https://yandex.kz/maps/-/CKA0jNoj) и т.д.).

---

In [21]:
# Создадим датафрейм с данными по сетевым заведениям:
chain_num_seats = data[data['chain'] == 'да'].copy()
# Попробуем применить минимальную предобработку:
chain_num_seats['object_name'] = chain_num_seats['object_name'].str.lower().str.replace('«', '').str.replace('»', '')
# Выведем его для просмотра:
chain_num_seats

id  object_name chain                        object_type  \
3       21894   пиццетория    да                               кафе   
6      148815     брусника    да                               кафе   
13      22579   алло пицца    да                               кафе   
16     144107     суши wok    да  предприятие быстрого обслуживания   
18      58565       тануки    да                           ресторан   
...       ...          ...   ...                                ...   
15361  208537  шоколадница    да                               кафе   
15362  209264  шоколадница    да                               кафе   
15363  209186  шоколадница    да                               кафе   
15364  221900  шоколадница    да                               кафе   
15365  222535  шоколадница    да                               кафе   

                                                 address  number  
3                город Москва, Абрамцевская улица, дом 1      40  
6          город Москва, переулок Сивцев Вражек, дом 6/2      10  
13                 город Москва, улица Авиаторов, дом 14      32  
16                   город Москва, Азовская улица, дом 3       7  
18     город Москва, Большая Академическая улица, дом 65     160  
...                                                  ...     ...  
15361       город Москва, 3-й Крутицкий переулок, дом 18      50  
15362           город Москва, улица Земляной Вал, дом 33      10  
15363           город Москва, улица Земляной Вал, дом 33      20  
15364  город Москва, поселение Московский, Киевское ш...      36  
15365             город Москва, Ходынский бульвар, дом 4      10  

[2968 rows x 6 columns]

Обработка будет состоять из двух этапов:

1. Привести все названия к нижнему регистру;
2. Изучить названия и привести названия сетей к единым названиям;
3. Заменить то что "пострадало" после массовой обработки.

In [22]:
chain_num_obj = chain_num_seats.copy()
chain_num_obj.rename(columns={'object_name': 'number_of_rests'})
chain_num_obj['oname'] = chain_num_obj['object_name']
chain_num_obj_grouped = chain_num_obj.groupby('oname', as_index=False).agg(
    {'object_name': 'count', 'number': 'mean', 'number':'median'})
chain_num_obj_grouped = chain_num_obj_grouped.rename(columns={'object_name':'number_of_rests'})
chain_num_obj_grouped

oname  number_of_rests  number
0            beverly hills diner                1    88.0
1                       bierloga                1    75.0
2                  black & white                1    40.0
3                     bocconcino                3    68.0
4                          boobo                1    46.0
..                           ...              ...     ...
549                    ямми микс                1     2.0
550                    ян примус                1   300.0
551  японский ресторан ваби-саби                1   155.0
552                        япоша                2   105.0
553                          ёрш                4   104.0

[554 rows x 3 columns]

In [23]:
# Построим scatter plot:
chain_num_obj_graph = px.scatter(chain_num_obj_grouped, x='number_of_rests', y='number')
chain_num_obj_graph.update_layout(
    title='Среднее количество посадочных мест в сетях',
    xaxis_title_text='Количество объектов',
    yaxis_title_text='Среднее количество посадочных мест в сети',
    template='plotly_dark'
    )
chain_num_obj_graph.show()

In [24]:
# Посмотрим на boxplot построенный по количеству посадочных мест в этих заведениях:
chain_num_obj_grouped_graph = go.Figure()
chain_num_obj_grouped_graph.add_trace(go.Box(
    y=chain_num_obj_grouped['number'],
    boxpoints='all',
    boxmean='sd'
))
chain_num_obj_grouped_graph.update_layout(
    title_text="Количество посадочных мест в сетях",
    xaxis=dict(showticklabels=False),
    xaxis_title_text='',
    template='plotly_dark'
    )

<a href="#up">Наверх</a>
<a name="s2_3"></a>
### Вывод - по типичному количеству посадочных мест
#### Больше всего представлены мелкие сети до 50 посадочных мест
Что характерно для сетевых заведений: много заведений с небольшим числом посадочных мест в каждом или мало заведений с большим количеством посадочных мест?
Диаграмма рассеяния указывает как и boxplot (для общего понимания картины, что для сетевых заведений г. Москвы характерное количество посадочных мест (около 50) с малым количеством заведений, множественное распространение с таким же (либо сопоставимым количеством мест) - это редкость, либо известные сети типа Макдоналс, KFC, Burger King.

---
<a href="#up">Наверх</a>
<a name="s2_5"></a>
### Количество посадочных мест по типам заведений
Для каждого вида объекта общественного питания опишите среднее количество посадочных мест. Какой вид предоставляет в среднем самое большое количество посадочных мест? Постройте графики.

In [25]:
type_num_seats_graph = go.Figure()

for i in chain_num_seats['object_type'].unique():
    type_num_seats_graph.add_trace(go.Box(
    y = chain_num_seats[chain_num_seats['object_type'] == i]['number'],
    name = i,
    boxmean='sd'
    ))

type_num_seats_graph.update_layout(
#    xaxis={'categoryorder':'total descending'}, 
    title='Среднее количество посадочных мест в сетях',
    xaxis_title_text='',
    yaxis_title_text='Количество посадочных мест',
    template='plotly_dark'
    )

type_num_seats_graph.show()

<a href="#up">Наверх</a>
<a name="s2_3"></a>
### Вывод - по типичному количеству посадочных мест по типам заведений
#### Больше всего посадочных мест в ресторанах

In [27]:
restaurants_median = chain_num_seats[chain_num_seats['object_type'] == 'ресторан']['number'].median()
restaurants_mean = chain_num_seats[chain_num_seats['object_type'] == 'ресторан']['number'].mean()
restaurants_sd = chain_num_seats[chain_num_seats['object_type'] == 'ресторан']['number'].std()

canteens_median = chain_num_seats[chain_num_seats['object_type'] == 'столовая']['number'].median()
canteens_mean = chain_num_seats[chain_num_seats['object_type'] == 'столовая']['number'].mean()
canteens_sd = chain_num_seats[chain_num_seats['object_type'] == 'столовая']['number'].std()

print(f"На box plot представлены все типы заведений. По описанию средних можно увидеть, \
что рестораны имеют наибольшее количество посадочных\ мест с медианой {round(restaurants_median,1)}, средним {round(restaurants_mean,1)}\
и стандартным отклонением ±{round(restaurants_sd, 1)} посадочных места. \
Может показаться, что у столовых медиана ({round(canteens_median,1)}) и среднее ({round(canteens_mean,1)}) \
больше, однако SD у этого типа гораздо меньше – ±{round(canteens_sd,1)}")

На box plot представлены все типы заведений. По описанию средних можно увидеть, что рестораны имеют наибольшее количество посадочных\ мест с медианой 88.5, средним 100.2и стандартным отклонением ±74.4 посадочных места. Может показаться, что у столовых медиана (120.0) и среднее (112.3) больше, однако SD у этого типа гораздо меньше – ±53.9


---
<a href="#up">Наверх</a>
<a name="s2_6"></a>
### Выделение улиц в столбце address
Выделите в отдельный столбец информацию об улице из столбца address.

In [28]:
# Исключим все поселения:
data_address = data[data.apply(lambda x: False if 'поселение' in x['address'] else True, axis=1)].copy()

In [29]:
# избавимся от "города Москва, ":
regex_pat = re.compile(r'город Москва, ', flags=re.IGNORECASE)
data_address['address'] = data['address'].str.replace(regex_pat, '')

# Добавим ту же самую предобработку данных с приведением к нижнему регистру и заменой кавычек:
data_address['object_name'] = data_address['object_name'].str.lower().str.replace('«', '').str.replace('»', '')

# Напишем функцию которая поможет 
def splt_comma(str_in, n_from, n_to):
    if n_to is None:
        return str(str_in.split(',')[n_from])
    
    if n_to is not None:
        return str_in.split(',')[n_from : n_to]
    
def street(string):
    if (('город' in string) or ('поселок' in string)) and not ('городская' in string):
        if len(splt_comma(string,-3,-1)) > 1: 
            return ','.join(splt_comma(string,-3,-1))
        else: 
            return splt_comma(string,0,None)
    if 'поселение' in string: 
        return splt_comma(string,-2,-1)
    else: 
        return splt_comma(string,0,None)

# Создадим столбец street:
data_address['street'] = data_address['address'].apply(street)

In [30]:
# Проверим рандомные 30 значений на адекватность названий улиц:
random_int = random.randint(0, data_address.shape[0])
data_address.sort_values(by='number')[random_int:random_int+30]

id                            object_name chain  \
12524  196092                       ресторан хэ цзин   нет   
10367  181501                               мольберт   нет   
1642    26644                    ресторан i like bar   нет   
6186   125150                             bocconcino    да   
9442   168375                                  italy   нет   
933     71429                                 тануки    да   
298     24240                            шоколадница    да   
598     26678                            шоколадница    да   
3405    71665                       кафе шоколадница    да   
13806   27044                               колтрейн   нет   
3560    25449                              starbucks    да   
12406  194009                  ресторан особое место   нет   
5415    23125                          кафе кофе тун    да   
6946    21187  буфет общества частных коневладельцев   нет   
5247    25916                       ресторан миш мэш   нет   
12110  182085                             додо пицца    да   
5282    82639                        крошка картошка    да   
5283    82640                             макдоналдс    да   
5281    21627                                теремок    да   
2626    22772                           старый город   нет   
6960   119614                            шоколадница    да   
7234   148928                             додо пицца    да   
3941   149336                          черный какаду   нет   
5864   124994                 ресторан своя компания   нет   
2365    27860                                  мнити   нет   
13959  205126                                 boroda   нет   
7228    25936                      ресторан белл паб   нет   
8076   112200                        бар хлеб и вино   нет   
2332    99154                            вино и мясо   нет   
12401  193058                              hang kong   нет   

                             object_type  \
12524                           ресторан   
10367                               кафе   
1642                            ресторан   
6186                            ресторан   
9442                            ресторан   
933                             ресторан   
298                                 кафе   
598                                 кафе   
3405                                кафе   
13806                           ресторан   
3560                                кафе   
12406                           ресторан   
5415                                кафе   
6946                               буфет   
5247                                кафе   
12110                               кафе   
5282   предприятие быстрого обслуживания   
5283   предприятие быстрого обслуживания   
5281                                кафе   
2626                                кафе   
6960                                кафе   
7234                                кафе   
3941                                кафе   
5864                            ресторан   
2365                            столовая   
13959                                бар   
7228                            ресторан   
8076                                 бар   
2332                            ресторан   
12401                               кафе   

                                                 address  number  \
12524                             улица Косыгина, дом 15      68   
10367                         Мантулинская улица, дом 12      68   
1642                           Марксистская улица, дом 7      68   
6186                         Ленинский проспект, дом 109      68   
9442                    Большая Грузинская улица, дом 69      68   
933                              улица Пришвина, дом 9/2      68   
298             улица Грузинский Вал, дом 26, строение 1      68   
598                        Ленинский проспект, дом 68/10      68   
3405              Мичуринский проспект, дом 22, корпус 1      68   
13806                             улица Кос

In [31]:
# Проверим чтобы поселения и города отображались как строки:
data_address[data_address['address'].apply(lambda x: True if 'город' in x and 'городская' not in x else False)].head(3)

id    object_name chain object_type  \
1386  22331  столовая миэт   нет    столовая   
1405  20199            это   нет        кафе   
1406  68113         френдс   нет        кафе   

                                     address  number  \
1386  город Зеленоград, улица Юности, дом 11      56   
1405            город Зеленоград, корпус 401      35   
1406            город Зеленоград, корпус 435      34   

                              street  
1386  город Зеленоград, улица Юности  
1405                город Зеленоград  
1406                город Зеленоград

In [32]:
# Проверим чтобы среди улиц также не было списков и они не сломали группировку:
data_address[data_address.apply(lambda x: True if isinstance(x['street'], list) else False, axis=1)]

Empty DataFrame
Columns: [id, object_name, chain, object_type, address, number, street]
Index: []

---
<a href="#up">Наверх</a>
<a name="s2_7_1"></a>
### Топ-10 улиц
Постройте график топ-10 улиц по количеству объектов общественного питания. Воспользуйтесь внешней информацией и ответьте на вопрос — в каких районах Москвы находятся эти улицы? 

In [33]:
# Соберём топ-10 улиц и отсортируем их по количеству объектов, при этом удалим все поселения и отдельные города, 
# так сравнивать небольшой город с улицей некорректно - масштаб этого несопоставим:

data_streets = data_address[data_address.apply(
    lambda x: False if ('город' in x['street']) or ('посёлок' in x['street']) else True, axis=1)].copy()

data_streets = data_streets.groupby('street', as_index=False).agg(
    {'id':'count'}).rename(columns={'id':'num_of_rests'})
top_10 = data_streets.sort_values(by='num_of_rests', ascending=False).head(10)
top_10

street  num_of_rests
1411           проспект Мира           204
1026       Профсоюзная улица           183
716   Ленинградский проспект           173
1011  Пресненская набережная           167
442         Варшавское шоссе           162
719       Ленинский проспект           148
1408    проспект Вернадского           132
707     Кутузовский проспект           114
634          Каширское шоссе           112
639     Кировоградская улица           110

In [34]:
address_registry = pd.read_csv('mos_ru_dataset.csv', sep=';', comment='#')
# Переименуем имеющиеся колонки:
address_registry = address_registry.rename(columns={
    'P1':'city', 'P5':'district', 'P7':'street', 'ADM_AREA': 'adm_area'})
# Сгруппируем по названию улицы:
address_registry = address_registry.groupby('street', as_index=False).first()
# и выведем его:
address_registry

street          city  \
0        1-й Автозаводский проезд  город Москва   
1         1-й Амбулаторный проезд  город Москва   
2     1-й Бабьегородский переулок  город Москва   
3         1-й Балтийский переулок  город Москва   
4          1-й Басманный переулок  город Москва   
...                           ...           ...   
4911                  улица Ясная  город Москва   
4912        улица местечко Барыши  город Москва   
4913                 шоссе Фрезер  город Москва   
4914            шоссе Энтузиастов  город Москва   
4915        южнее деревни Внуково  город Москва   

                                district  \
0        муниципальный округ Даниловский   
1           муниципальный округ Аэропорт   
2           муниципальный округ Якиманка   
3           муниципальный округ Аэропорт   
4     муниципальный округ Красносельский   
...                                  ...   
4911                                None   
4912                                None   
4913   муниципальный округ Нижегородский   
4914          муниципальный округ Перово   
4915                                None   

                                   adm_area  
0              Южный административный округ  
1           Северный административный округ  
2        Центральный административный округ  
3           Северный административный округ  
4        Центральный административный округ  
...                                     ...  
4911        Троицкий административный округ  
4912  Новомосковский административный округ  
4913   Юго-Восточный административный округ  
4914       Восточный административный округ  
4915  Новомосковский административный округ  

[4916 rows x 4 columns]

In [35]:
# Присоединим районы и округа к топ-10 улиц:
top_10_joined = top_10.merge(address_registry, on='street').drop(columns='city')
top_10_joined

street  num_of_rests  \
0           проспект Мира           204   
1       Профсоюзная улица           183   
2  Ленинградский проспект           173   
3  Пресненская набережная           167   
4        Варшавское шоссе           162   
5      Ленинский проспект           148   
6    проспект Вернадского           132   
7    Кутузовский проспект           114   
8         Каширское шоссе           112   
9    Кировоградская улица           110   

                                   district  \
0          муниципальный округ Алексеевский   
1               муниципальный округ Ясенево   
2                 муниципальный округ Сокол   
3           муниципальный округ Пресненский   
4       муниципальный округ Чертаново Южное   
5           муниципальный округ Гагаринский   
6  муниципальный округ Проспект Вернадского   
7          муниципальный округ Дорогомилово   
8    муниципальный округ Нагатино-Садовники   
9       муниципальный округ Чертаново Южное   

                                  adm_area  
0  Северо-Восточный административный округ  
1      Юго-Западный административный округ  
2          Северный административный округ  
3       Центральный административный округ  
4             Южный административный округ  
5      Юго-Западный административный округ  
6          Западный административный округ  
7          Западный административный округ  
8             Южный административный округ  
9             Южный административный округ

In [36]:
top_10_grouped = top_10_joined.groupby('adm_area').agg({'num_of_rests':'sum'}).sort_values('num_of_rests', ascending=False).copy()
top_10_grouped

num_of_rests
adm_area                                             
Южный административный округ                      384
Юго-Западный административный округ               331
Западный административный округ                   246
Северо-Восточный административный округ           204
Северный административный округ                   173
Центральный административный округ                167

In [34]:
# Много это или мало? Добавим другие показатели:
data_population = pd.read_csv('moscow_population_density.csv', header=1)
data_population = data_population.rename(
    columns={"Адми-\nнистра-\nтивный\nокруг":'adm_area', 
             "Пло-\nщадь,\n[6][7]\nкм²":'area', 'Насе-\nление\n2019[8],\nчел.':'population'})

# Сделаем небольшую предобработку данных: 
data_population_clean = data_population[['adm_area','area','population']].copy()
data_population_clean['population'] = data_population_clean.apply(
    lambda x: re.sub(r'[↗|↘| ]', '', x['population']), axis=1).copy().astype('int')
data_population_clean['area'] = data_population_clean.apply(
    lambda x: re.sub(r'[,]', '.', x['area']), axis=1).copy().astype('float').copy()

# Добавим столбец по плотности населения:
data_population_clean_grouped = data_population_clean.groupby('adm_area', as_index=False).sum()
data_population_clean_grouped['population_density'] = data_population_clean_grouped['population']\
                                                            /data_population_clean_grouped['area']

# Переделаем названия чтобы они соответствовали сокращённым:
adm_district_list = address_registry['adm_area'].unique()
data_population_clean_grouped.sort_values('population_density')

# Добавим соответствия для длинных имён (к сожалению, руками):
def rename_adm_district(short_name):
    if short_name['adm_area'] == 'ТАО': return 'Троицкий административный округ'  
    if short_name['adm_area'] == 'НАО': return 'Новомосковский административный округ'
    if short_name['adm_area'] == 'ЗелАО': return 'Зеленоградский административный округ'
    if short_name['adm_area'] == 'ЗАО': return 'Западный административный округ'
    if short_name['adm_area'] == 'САО': return 'Северный административный округ'
    if short_name['adm_area'] == 'СВАО': return 'Северо-Восточный административный округ'
    if short_name['adm_area'] == 'СЗАО': return 'Северо-Западный административный округ'
    if short_name['adm_area'] == 'ЦАО': return 'Центральный административный округ'
    if short_name['adm_area'] == 'ЮАО': return 'Южный административный округ'
    if short_name['adm_area'] == 'ЮВАО': return 'Юго-Восточный административный округ'
    if short_name['adm_area'] == 'ЮЗАО': return 'Юго-Западный административный округ'
    if short_name['adm_area'] == 'ВАО': return 'Восточный административный округ'
    
# Применим функцию:
data_population_clean_grouped['adm_area'] = data_population_clean_grouped.apply(
                                                            rename_adm_district, axis=1)
data_population_clean_grouped

adm_area     area  population  \
0          Восточный административный округ   156.70     1523420   
1           Западный административный округ   153.05     1392993   
2     Зеленоградский административный округ    37.21      246535   
3     Новомосковский административный округ   361.12      259682   
4           Северный административный округ   115.19     1185416   
5   Северо-Восточный административный округ   102.49     1431976   
6    Северо-Западный административный округ    92.52     1008968   
7           Троицкий административный округ  1026.34      127349   
8        Центральный административный округ    66.22      782048   
9              Южный административный округ   130.65     1793456   
10     Юго-Восточный административный округ   117.54     1418430   
11      Юго-Западный административный округ   110.41     1445006   

    population_density  
0          9721.888960  
1          9101.555047  
2          6625.503897  
3           719.101684  
4         10290.962757  
5         13971.860669  
6         10905.404237  
7           124.080714  
8         11809.845968  
9         13727.179487  
10        12067.636549  
11        13087.636989

In [35]:
# Соберём датафреймы с плотностью населения:
top_10_joined_final = top_10_joined.merge(
    data_population_clean_grouped, on='adm_area').round(1).sort_values('num_of_rests', ascending=False)
top_10_joined_final

street  num_of_rests  \
0           проспект Мира           204   
1       Профсоюзная улица           183   
3  Ленинградский проспект           173   
4  Пресненская набережная           167   
5        Варшавское шоссе           162   
2      Ленинский проспект           148   
8    проспект Вернадского           132   
9    Кутузовский проспект           114   
6         Каширское шоссе           112   
7    Кировоградская улица           110   

                                   district  \
0          муниципальный округ Алексеевский   
1               муниципальный округ Ясенево   
3                 муниципальный округ Сокол   
4           муниципальный округ Пресненский   
5       муниципальный округ Чертаново Южное   
2           муниципальный округ Гагаринский   
8  муниципальный округ Проспект Вернадского   
9          муниципальный округ Дорогомилово   
6    муниципальный округ Нагатино-Садовники   
7       муниципальный округ Чертаново Южное   

                                  adm_area   area  population  \
0  Северо-Восточный административный округ  102.5     1431976   
1      Юго-Западный административный округ  110.4     1445006   
3          Северный административный округ  115.2     1185416   
4       Центральный административный округ   66.2      782048   
5             Южный административный округ  130.6     1793456   
2      Юго-Западный административный округ  110.4     1445006   
8          Западный административный округ  153.0     1392993   
9          Западный административный округ  153.0     1392993   
6             Южный административный округ  130.6     1793456   
7             Южный административный округ  130.6     1793456   

   population_density  
0             13971.9  
1             13087.6  
3             10291.0  
4             11809.8  
5             13727.2  
2             13087.6  
8              9101.6  
9              9101.6  
6             13727.2  
7             13727.2

In [36]:
top_10_grouped_w_population = top_10_grouped.merge(data_population_clean_grouped, on='adm_area').round(1)
# Добавим соотношение количества людей на 
top_10_grouped_w_population['num_per_person'] = (
    top_10_grouped_w_population['population'] / top_10_grouped_w_population['num_of_rests']
    ).round(1)
# Выведем его:
top_10_grouped_w_population

adm_area  num_of_rests   area  population  \
0             Южный административный округ           384  130.6     1793456   
1      Юго-Западный административный округ           331  110.4     1445006   
2          Западный административный округ           246  153.0     1392993   
3  Северо-Восточный административный округ           204  102.5     1431976   
4          Северный административный округ           173  115.2     1185416   
5       Центральный административный округ           167   66.2      782048   

   population_density  num_per_person  
0             13727.2          4670.5  
1             13087.6          4365.6  
2              9101.6          5662.6  
3             13971.9          7019.5  
4             10291.0          6852.1  
5             11809.8          4682.9

<a href="#up">Наверх</a>
<a name="s2_7_2"></a>
### Вывод - по топ-10 улиц с наибольшим количеством заведений
#### Топ-3 - это юг, юго-запад и запад
Сгруппировав улицы с наибольшим количеством заведений мы получили 384 заведений всех типов. Интересно было соотнести количество заведений и количество населения, что показало, что самая высокая численность населения - и самое большое количество заведений - в ЮАО. Также мы рассчитали отношение количества населения к населению. Самый высокий коэффициент вышел у СВАО и САО.

---
<a href="#up">Наверх</a>
<a name="s2_7_3"></a>
### Топ-10 улиц - внешние источники и ссылки.
Для исследования использовался [датасет адресного реестра города Москвы](https://data.mos.ru/opendata/7705031674-adresniy-reestr-zdaniy-i-soorujeniy-v-gorode-moskve) Открытого портала данных и [Википедию](https://ru.wikipedia.org/wiki/Список_районов_и_поселений_Москвы) (перепись населения г. Москвы на 2019 год).

---
<a href="#up">Наверх</a>
<a name="s2_8_1"></a>
### Число улиц с одним объектом общественного питания
Найдите число улиц с одним объектом общественного питания. Воспользуйтесь внешней информацией и ответьте на вопрос — в каких районах Москвы находятся эти улицы?

In [37]:
# Внешняя информация уже подгружена, поэтому остается только собрать датафрейм и отфильтровать:
bottom_1 = data_streets[data_streets['num_of_rests'] == 1]
bottom_1_joined = bottom_1.merge(address_registry, on='street')
# В этих районах хуже всего с заведениями и практически нет конкуренции. Возможно, там действительно нечего ловить:
bottom_1_joined.groupby('district', as_index=False).first()

district  \
0    муниципальное образование Савелки   
1     муниципальный округ Алексеевский   
2     муниципальный округ Алтуфьевский   
3            муниципальный округ Арбат   
4         муниципальный округ Аэропорт   
..                                 ...   
99    муниципальный округ Южное Тушино   
100   муниципальный округ Южнопортовый   
101       муниципальный округ Якиманка   
102    муниципальный округ Ярославский   
103        муниципальный округ Ясенево   

                                            street  num_of_rests  \
0                                   Сосновая аллея             1   
1                             1-й Рижский переулок             1   
2                                Стандартная улица             1   
3                     1-й Николощеповский переулок             1   
4                          1-й Балтийский переулок             1   
..                                             ...           ...   
99                               Нелидовская улица             1   
100                         3-й Кожуховский проезд             1   
101                      1-й Голутвинский переулок             1   
102                                Лосевская улица             1   
103  38-й километр Московской Кольцевой Автодороги             1   

             city                                 adm_area  
0    город Москва    Зеленоградский административный округ  
1    город Москва  Северо-Восточный административный округ  
2    город Москва  Северо-Восточный административный округ  
3    город Москва       Центральный административный округ  
4    город Москва          Северный административный округ  
..            ...                                      ...  
99   город Москва   Северо-Западный административный округ  
100  город Москва     Юго-Восточный административный округ  
101  город Москва       Центральный административный округ  
102  город Москва  Северо-Восточный административный округ  
103  город Москва      Юго-Западный административный округ  

[104 rows x 5 columns]

Больше сотни улиц, где всего одно кафе.

In [38]:
data_streets.groupby(
    'street', as_index=False).agg({'num_of_rests':'count'}).merge(
        address_registry, on='street')

street  num_of_rests          city  \
0     1-й Автозаводский проезд             1  город Москва   
1      1-й Амбулаторный проезд             1  город Москва   
2      1-й Балтийский переулок             1  город Москва   
3       1-й Басманный переулок             1  город Москва   
4      1-й Ботанический проезд             1  город Москва   
...                        ...           ...           ...   
1813              улица Юности             1  город Москва   
1814       улица Юных Ленинцев             1  город Москва   
1815           улица Яблочкова             1  город Москва   
1816              шоссе Фрезер             1  город Москва   
1817         шоссе Энтузиастов             1  город Москва   

                                district  \
0        муниципальный округ Даниловский   
1           муниципальный округ Аэропорт   
2           муниципальный округ Аэропорт   
3     муниципальный округ Красносельский   
4           муниципальный округ Свиблово   
...                                  ...   
1813         муниципальный округ Вешняки   
1814    муниципальный округ Текстильщики   
1815       муниципальный округ Бутырский   
1816   муниципальный округ Нижегородский   
1817          муниципальный округ Перово   

                                     adm_area  
0                Южный административный округ  
1             Северный административный округ  
2             Северный административный округ  
3          Центральный административный округ  
4     Северо-Восточный административный округ  
...                                       ...  
1813         Восточный административный округ  
1814     Юго-Восточный административный округ  
1815  Северо-Восточный административный округ  
1816     Юго-Восточный административный округ  
1817         Восточный административный округ  

[1818 rows x 5 columns]

In [39]:
# Создадим dataframe сгруппировав датафрейм по улицам всего с одним заведением и присоединим таблицу с административными округами: 
bottom_1_by_adm_district = bottom_1_joined.groupby(
    'adm_area', as_index=False).agg({'street':'count'}).sort_values('street', ascending=False)
bottom_1_by_adm_district = bottom_1_by_adm_district.rename(columns={'street':'num_of_rests'})
# Отсортируем:
bottom_1_by_adm_district_grouped = bottom_1_by_adm_district.groupby(
    'adm_area', as_index=False).sum().sort_values('num_of_rests',ascending=False)

# Построим график:
bottom_1_by_adm_district_grouped_graph = px.bar(
    bottom_1_by_adm_district, 
    x='num_of_rests', 
    y='adm_area', 
    text='num_of_rests', 
    orientation='h'
    )

bottom_1_by_adm_district_grouped_graph.update_yaxes(autorange="reversed")
bottom_1_by_adm_district_grouped_graph.update_traces(textposition='inside')
bottom_1_by_adm_district_grouped_graph.update_layout(
    template='plotly_dark',
    title_text='Количество улиц с 1 заведением',
    xaxis_title='Количество заведений',
    yaxis_title='',
    )

bottom_1_by_adm_district_grouped_graph.show()

In [40]:
# Посчитаем общее количество таких заведений:
print(f"Всего {bottom_1_by_adm_district['num_of_rests'].sum()} улиц с 1 заведением.")

Всего 536 улиц с 1 заведением.


<a href="#up">Наверх</a>
<a name="s2_8_2"></a>
### Вывод по количеству улиц с одним заведением
#### Топ-3 округов по количеству улиц с 1 заведением - ЦАО, СВАО и ВАО.
Улиц с 1 заведением целых 536. Примечательно, улицы с одним заведением есть в каждом из административных округов. Не все заведения подходят для массового распространения.

---
<a href="#up">Наверх</a>
<a name="s2_9_1"></a>
### Распределение количества посадочных мест для улиц с большим количеством объектов
Посмотрите на распределение количества посадочных мест для улиц с большим количеством объектов общественного питания. Какие закономерности можно выявить?

Определим улицы с "большим количеством объектов" как верхний квартиль.

In [41]:
data_top_addresses = data_address.groupby(
    ['street','chain','object_type'],
    as_index=False).agg(
        {'number':'median', 'id':'count'}
            ).rename(
                columns={'number':'number_of_seats','id':'number_of_rests'}
                    ).sort_values('number_of_rests', ascending=False).reset_index(drop=True).round(1)

data_top_addresses

street chain object_type  number_of_seats  \
0       Пресненская набережная   нет        кафе             24.0   
1                проспект Мира   нет        кафе             45.0   
2            Профсоюзная улица   нет        кафе             24.0   
3      улица Ленинская Слобода   нет        кафе             10.0   
4             город Зеленоград   нет    столовая            160.0   
...                        ...   ...         ...              ...   
5806      Новомосковская улица   нет         бар             48.0   
5807      Новомосковская улица   нет       буфет             60.0   
5808     Новоостаповская улица   нет    столовая             48.0   
5809  Новопеределкинская улица    да        кафе             45.0   
5810            Пушечная улица   нет         бар             40.0   

      number_of_rests  
0                  78  
1                  73  
2                  59  
3                  56  
4                  54  
...               ...  
5806                1  
5807                1  
5808                1  
5809                1  
5810                1  

[5811 rows x 5 columns]

In [42]:
data_top_addresses_description = data_top_addresses.describe().round(1)
data_top_addresses_description

number_of_seats  number_of_rests
count           5811.0           5811.0
mean              59.6              2.6
std               67.8              4.1
min                0.0              1.0
25%               15.0              1.0
50%               40.0              1.0
75%               80.0              3.0
max             1400.0             78.0

In [43]:
data_top_addresses_top = data_top_addresses[:int(data_top_addresses_description['number_of_rests']['count']*0.25)]
data_top_addresses_top.head(10)

street chain object_type  number_of_seats  number_of_rests
0   Пресненская набережная   нет        кафе             24.0               78
1            проспект Мира   нет        кафе             45.0               73
2        Профсоюзная улица   нет        кафе             24.0               59
3  улица Ленинская Слобода   нет        кафе             10.0               56
4         город Зеленоград   нет    столовая            160.0               54
5      улица Сущёвский Вал   нет        кафе             13.0               53
6         Варшавское шоссе   нет        кафе             28.5               48
7       Ленинский проспект   нет        кафе             32.0               46
8     улица Миклухо-Маклая   нет        кафе             10.0               45
9        Ходынский бульвар   нет        кафе              0.0               44

In [44]:
data_top_addresses_graph = px.scatter(
    data_top_addresses_top, 
    x="number_of_seats", 
    y="number_of_rests", 
    color='chain',
    opacity=0.5
    )
data_top_addresses_graph.update_layout(
    title_text='Количество посадочных мест и заведений в сетевых и несетевых заведениях',
    xaxis_title='Количество посадочных мест',
    yaxis_title='Количество заведений',
    template='plotly_dark')
data_top_addresses_graph.show()

In [45]:
data_top_addresses_graph = px.scatter(
    data_top_addresses, 
    x="number_of_seats", 
    y="number_of_rests", 
    color='object_type',
    opacity=0.6)
data_top_addresses_graph.update_layout(
    title_text='Количество посадочных мест и заведений по типу объекта',
    xaxis_title='Количество посадочных мест',
    yaxis_title='Количество заведений',
    template='plotly_dark')
data_top_addresses_graph.show()

<a href="#up">Наверх</a>
<a name="s2_9_2"></a>
### Вывод по распределению количества посадочных мест для улиц с большим количеством объектов
#### Чем больше в заведении посадочных мест, тем больше оно представлено на загруженных улицах
- Из-за ограниченного пространства много заведений с небольшим (до 100) количеством посадочных мест, кроме ресторанов и столовых
- Сетевые заведения представлены на загруженных улицах меньше, чем несетевые, что подтверждается и предыдущими графиками. 
- Кафе и предприятия быстрого питания также можно выделить в отдельную категорию - эти заведения с небольшим (до 60) количеством посадочных мест являются наиболее представленными.
- Столовые и рестораны можно выделить в отдельную категорию с наибольшим количестсвом мест - до 250, но с наименьшим распространением.
- Наименее интересные относительно немногочисленные отделы кулинарии, буфеты, закусочные и кафетерии - они плохо представлены и редки на загруженных улицах.

---
<a href="#up">Наверх</a>
<a name="s2_10"></a>
# Общий вывод
### Рекомендации
- __Тип заведения:__ кафе или предприятие быстрого питания, где можно использовать автоматизацию в виде роботов.
- __Количество посадочных мест:__ от 40 до 50.
- __Район расположения:__ для дальнейшего исследования (с учётом потенциального развития сети) следует рассматривать юг, юго-запад и запад. Также по расчёту 

__Прокомментируйте возможность развития сети:__ Предприятия быстрого питания и кафе являются самыми перспективными для развития сети.